packages/bridge-ui-v2/src/components/Bridge/Bridge.svelte

```diff
@@ -35,35 +49,205 @@
       $destNetwork = null;
     }
   }
+
+  async function approve() {
+    if (!$selectedToken || !$network || !$destNetwork) return;
+
+    const erc20Bridge = bridges.ERC20 as ERC20Bridge;
+
+    try {
+      const walletClient = await getConnectedWallet($network.id);
+
+      const tokenAddress = await getAddress({
+        token: $selectedToken,
+        srcChainId: $network.id,
+        destChainId: $destNetwork.id,
+      });
+
+      if (!tokenAddress) {
+        throw new Error('token address not found');
+      }
+
+      const spenderAddress = chainContractsMap[$network.id].tokenVaultAddress;
+
+      const txHash = await erc20Bridge.approve({
+        tokenAddress,
+        spenderAddress,
+        amount: $enteredAmount,
+        wallet: walletClient,
+      });
+
+      infoToast(
+        $t('bridge.approve.tx', {
+          values: {
+            token: $selectedToken.symbol,
+            url: `${PUBLIC_L1_EXPLORER_URL}/tx/${txHash}`,
+          },
+        }),
+      );
+
+      await pendingTransactions.add(txHash, $network.id);
+
+      successToast(
+        $t('bridge.approve.success', {
+          values: {
+            token: $selectedToken.symbol,
+          },
+        }),
+      );
+
+      // Let's run the validation again, which will update UI
+      amountComponent.validateAmount();
+    } catch (err) {
+      console.error(err);
+
+      switch (true) {
+        case err instanceof UserRejectedRequestError:
+          warningToast($t('bridge.approve.rejected'));
+          break;
+        case err instanceof NoAllowanceRequiredError:
+          errorToast($t('bridge.approve.no_allowance_required'));
+          break;
+        case err instanceof ApproveError:
+          // TODO: see contract for all possible errors
+          errorToast($t('bridge.approve.error'));
+          break;
+        default:
+          errorToast($t('bridge.approve.unknown_error'));
+      }
+    }
+  }
+
+  async function bridge() {
+    if (!$selectedToken || !$network || !$destNetwork || !$account?.address) return;
+
+    try {
+      const walletClient = await getConnectedWallet($network.id);
+
+      let bridge: Bridge;
+
+      // Common arguments for both ETH and ERC20 bridges
+      let bridgeArgs = {
+        to: $recipientAddress || $account.address,
+        wallet: walletClient,
+        srcChainId: $network.id,
+        destChainId: $destNetwork.id,
+        amount: $enteredAmount,
+        processingFee: $processingFee,
+      } as BridgeArgs;
+
+      if (isETH($selectedToken)) {
+        bridge = bridges.ETH as ETHBridge;
```

Comment:
this kind of logic looks like its getting a bit hairy, should this not be a reactive statement above, and Token simply have a domain `TokenType` property?


and the domain token type can be of type `ETH, ERC721, ERC1155, ERC20`?

```ts
$ : bridge = (token) => {
    return tokenTypeToBridgeType(token.tokenType);
}($selectedToken)

function tokenTypeToBridgeType(t: TokenType) {
   switch(t) {
      case TokenType.ETH:
      return bridges.ETH;
     case TokenType.ERC20:
     return bridges.ERC20;
    case TokenType.ERC721:
    return bridges.ERC721;
     case TokenType.ERC1155:
   return bridges.ERC1155
   }
}
```

and then similarily:

```ts
function tokenTypeToBridgeArgs(t: TokenType) {
  let args: BridgeArgs =  {
        to: $recipientAddress || $account.address,
        wallet: walletClient,
        srcChainId: $network.id,
        destChainId: $destNetwork.id,
        amount: $enteredAmount,
        processingFee: $processingFee,
      };
   switch(t) {
    case TokenType.ETH:
      return { ...bridgeArgs, bridgeAddress };
     case TokenType.ERC20:
     return  const tokenAddress = await getAddress({
          token: $selectedToken,
          srcChainId: $network.id,
          destChainId: $destNetwork.id,
        });
        if (!tokenAddress) {
          throw new Error('token address not found');
        }
        const tokenVaultAddress = chainContractsMap[$network.id].tokenVaultAddress;
        const isTokenAlreadyDeployed = await isDeployedCrossChain({
          token: $selectedToken,
          srcChainId: $network.id,
          destChainId: $destNetwork.id,
        });
        bridgeArgs = {
          ...bridgeArgs,
          tokenAddress,
          tokenVaultAddress,
          isTokenAlreadyDeployed,
        };
        
      // etc, for erc721/1155
   }
}
```

and then your in-form bridge function becomes, without changing for additioonal future bridge functions:

```ts
     bridge.bridge(tokenTypeToBridgeArgs($selectedToken.tokenType));
```

and does not change for any future iterations of ERC1155,ERC721, or anything else we add in the future, the logic is encapsulated within domain helpers.

Reply:
Agree, was a temporal approach  when I started until I figure out NFT part, then I never changed it. I'll include the type in to token.

Reply:
Change made. Token contains now the type